### ISSUE TITLE
FocusNode and TextEditingController not disposed

### ISSUE THREAD

**Issue by bawelter** [jifalops/datetime_picker_formfield#55]
_DateTimeFieldState class doesn't call the dispose method if _focusNode / _controller has been defined.

**Comment by jifalops** [comment]
I think this is the correct behavior as it's exactly how `TextFormField` does it. Only the controllers/focusNodes that are defined on the widget should be disposed. If the state had to define its own, they don't need disposed. 

Is there a related issue you've noticed?

**Comment by bawelter** [comment]
I believe if it has a listener registered to it you need to either need to unregister the listener or dispose of it otherwise you may have a potential memory leak.  I may be wrong though, I was just following the docs and my experience with memory management.

Let me know what you find out or if you have any questions.  I will take a look at the TextFormField when I get a chance.

**Comment by jifalops** [comment]
I'll take another look at it. I thought listeners were removed and added when the state's widget changes.

**Comment by bawelter** [comment]
So I looked at their code and it is fine because they only add the listener to the controller passed in by the user and then they remove it in the dispose method or if it changes.  Every time a listener is added it needs to be removed when your done with it.  The system doesn’t know how to do this automatically so they have dispose methods which are called when the widget is removed from the tree and won’t be used again.  If you don’t remove the listeners / call dispose then the resources won’t be released and the system wont be able to to garbage collect them and cause a memory leak.  That’s if I understand everything correctly.

In this case the difference is that you add listeners to the controller / focus node if one isn’t provided and then never remove it in the dispose method.  They don’t add the listener when the controller isn’t provided.

**Comment by bawelter** [comment]
I created a pull request with a suggested fix for this.  Let me know if you have any questions.

This pull request also adds the listener to the controller / focus node and disposes of them properly in the didUpdateWidget.  This would have caused an issue where sometimes randomly the date time picker would not pop up or the onChanged function would stop firing.

**Comment by VictorUvarov** [comment]
I'm getting this error when using the widget. I assume it is related to this issue. How soon can we expect to see this resolved?
```
E/flutter (31822): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: Concurrent modification during iteration: Instance of '_CompactLinkedHashSet<FocusNode>'.
E/flutter (31822): #0      _CompactIterator.moveNext  (dart:collection-patch/compact_hash.dart:443:7)
E/flutter (31822): #1      FocusManager._applyFocusChange 
package:flutter/…/widgets/focus_manager.dart:1056
E/flutter (31822): #2      _rootRun  (dart:async/zone.dart:1120:38)
E/flutter (31822): #3      _CustomZone.run  (dart:async/zone.dart:1021:19)
E/flutter (31822): #4      _CustomZone.runGuarded  (dart:async/zone.dart:923:7)
E/flutter (31822): #5      _CustomZone.bindCallbackGuarded.<anonymous closure>  (dart:async/zone.dart:963:23)
E/flutter (31822): #6      _rootRun  (dart:async/zone.dart:1124:13)
E/flutter (31822): #7      _CustomZone.run  (dart:async/zone.dart:1021:19)
E/flutter (31822): #8      _CustomZone.runGuarded  (dart:async/zone.dart:923:7)
E/flutter (31822): #9      _CustomZone.bindCallbackGuarded.<anonymous closure>  (dart:async/zone.dart:963:23)
E/flutter (31822): #10     _microtaskLoop  (dart:async/schedule_microtask.dart:41:21)
E/flutter (31822): #11     _startMicrotaskLoop  (dart:async/schedule_microtask.dart:50:5)
E/flutter (31822):
```

**Comment by jifalops** [comment]
@VictorUvarov that is unrelated to this issue and is instead related to hiding the keyboard. See #51. Which version are you using?

**Comment by VictorUvarov** [comment]
I'm using the newest version. The widget is inside of a form. This is the widget it is wrapped inside of.
```dart
class BasicDateTimeField extends StatelessWidget {
  final Function onValidate;
  final TextEditingController controller;

  const BasicDateTimeField({
    Key key,
    @required this.onValidate,
    @required this.controller,
  })  : assert(onValidate != null),
        assert(controller != null),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      children: <Widget>[
        DateTimeField(
          validator: onValidate,
          controller: controller,
          decoration: InputDecoration(
            hintText: "Select a Date and Time",
            border: OutlineInputBorder(
              borderRadius: BorderRadius.circular(UiSettings.textFieldRounding),
            ),
          ),
          //2012-02-27 13:27:00
          format: DateFormat('yyyy-MM-dd hh:mm:ss'),
          onShowPicker: (context, currentValue) async {
            final date = await showDatePicker(
              context: context,
              firstDate: DateTime(1900),
              initialDate: currentValue ?? DateTime.now(),
              lastDate: DateTime(2100),
            );
            if (date != null) {
              final time = await showTimePicker(
                context: context,
                initialTime: TimeOfDay.fromDateTime(
                  currentValue ?? DateTime.now(),
                ),
              );
              return DateTimeField.combine(date, time);
            } else {
              return currentValue;
            }
          },
        ),
      ],
    );
  }
}
```

**Comment by jifalops** [comment]
@VictorUvarov when does the error happen? When you click the reset icon?

**Comment by VictorUvarov** [comment]
It happens immediately when I tap the text field for the date picker and immediately again when the time picker shows up.

**Comment by jifalops** [comment]
@VictorUvarov I've tried your code and am unable to reproduce the issue. Are you sure you're on v0.4.1 or v1.0.0-pre.2?

**Comment by bawelter** [comment]
We would need to look at your stateful widget where the TextEditingController is defined to identify the issue.

**Comment by jifalops** [comment]
Closing as this just landed in v0.4.2 and 1.0.0-pre.3. I'll keep an eye on the conversation here.
